processor/src/main/java/org/mapstruct/ap/internal/processor/MapperAnnotatedFormattingMessenger.java

```diff
@@ -0,0 +1,137 @@
+import javax.lang.model.element.VariableElement;
+import javax.tools.Diagnostic.Kind;
+
+import org.mapstruct.ap.internal.util.FormattingMessager;
+import org.mapstruct.ap.internal.util.Message;
+import org.mapstruct.ap.internal.util.TypeUtils;
+
+/**
+ * Handles redirection of errors/warnings so that they're shown on the mapper instead of hidden on a superclass.
+ *
+ * @author Ben Zegveld
+ */
+public class MapperAnnotatedFormattingMessenger implements FormattingMessager {
+
+    private FormattingMessager delegateMessager;
+    private TypeElement mapperTypeElement;
+    private TypeUtils typeUtils;
+
+    public MapperAnnotatedFormattingMessenger(FormattingMessager delegateMessager, TypeElement mapperTypeElement,
+                                              TypeUtils typeUtils) {
+        this.delegateMessager = delegateMessager;
+        this.mapperTypeElement = mapperTypeElement;
+        this.typeUtils = typeUtils;
+    }
+
+    @Override
+    public void printMessage(Message msg, Object... args) {
+        delegateMessager.printMessage( msg, args );
+    }
+
+    @Override
+    public void printMessage(Element e, Message msg, Object... args) {
+        delegateMessager
+                        .printMessage(
+                            determineDelegationElement( e ),
+                            determineDelegationMessage( e, msg ),
+                            determineDelegationArguments( e, msg, args ) );
+    }
+
+    @Override
+    public void printMessage(Element e, AnnotationMirror a, Message msg, Object... args) {
+        delegateMessager
+                        .printMessage(
+                            determineDelegationElement( e ),
+                            a,
+                            determineDelegationMessage( e, msg ),
+                            determineDelegationArguments( e, msg, args ) );
+    }
+
+    @Override
+    public void printMessage(Element e, AnnotationMirror a, AnnotationValue v, Message msg, Object... args) {
+        delegateMessager
+                        .printMessage(
+                            determineDelegationElement( e ),
+                            a,
+                            v,
+                            determineDelegationMessage( e, msg ),
+                            determineDelegationArguments( e, msg, args ) );
+    }
+
+    @Override
+    public void note(int level, Message log, Object... args) {
+        delegateMessager.note( level, log, args );
+    }
+
+    @Override
+    public boolean isErroneous() {
+        return delegateMessager.isErroneous();
+    }
+
+    private Object[] determineDelegationArguments(Element e, Message msg, Object[] args) {
+        if ( methodInMapperClass( e ) ) {
+            return args;
+        }
+        String originalMessage = String.format( msg.getDescription(), args );
+        return new Object[] { originalMessage, constructMethod( e ), e.getEnclosingElement().getSimpleName() };
```

Comment:
Instead of using `getSimpleName` I would suggest to use the `getQualifiedName` from `QualifiedNameable` if the `e.getEnclosingElement()` returns a `QualifiedNameable` (which it most likely does always here). 

I think this would make it clear in which type this happened.

Reply:
Having a qualified name is not really necessary here I think. You are referencing that class from the code there anyway, since you are either extending or using it.
I also believe it will obscure the source because of message length if we do it like that.

For example:
`This is the original message with an error. Occured at TargetType mapMethod(SourceType source) in GeneralMapper.`
vs
`This is the original message with an error. Occured at TargetType mapMethod(SourceType source) in org.mapstruct.business.component.product.storage.domain.general.GeneralMapper.`
(used a location that has a similar structure as in one of the product that I work on)

It should be clear which `GeneralMapper` it is, because the error appears near the line:
`public abstract class MySubDomainMapper extends GeneralMapper`

Although if you have multiple layers of hierarchy you might not see it immediate, but you can just click into the parent class to find out who that one is extending:
`public abstract class MySubDomainMapper extends MyDomainMapper`
`public abstract class MyDomainMapper extends GeneralMapper`

I'm not sure if the uses mappers that are not annotated with `@Mapper` or `@Mapping` will keep their errors with this change. It might even be that those will get their errors redirected here, come to think of it. But since you are referencing them explicitly with uses you should be able to find them with just the class name.

If on the other hand it is standard behavior for compiler messages to refer to the qualified name instead, then I agree with that it should become the qualified name. I just don't have that much knowledge yet about standard behavior for compiler/annotation processing.

Reply:
You are absolutely right. Having the simple name here makes absolutely more sense. I didn't thought about the fact that you already have the name in the location of the main mapper. Let's leave it as is